[Inductor CUTLASS backend] Step 4: CUDA (template) kernels

ipiszy opened the pull request:
This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* __->__ #107931
* #107901
* #107847
* #107802



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
[Inductor CUTLASS backend] Step 4: CUDA (template) kernels

[ghstack-poisoned]

kadeng commented on torch/_inductor/codegen/common.py:
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
+        return None
+
+
+class ChoiceCaller:
```
A small doc comment would be great: What's the purpose of this class / which problem does it solve? Is it supposed to have subclasses?

kadeng commented on torch/_inductor/codegen/common.py:
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
+
+class KernelTemplate:
+    """
+    Base class for defining kernel templates.
```
Which kind of kernel templates? ( e.g. Triton / C++ / Cutlass / any involving Jinja templates )

ipiszy replied:
Added some comments.

kadeng commented on torch/_inductor/codegen/common.py:
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
+    def __init__(self, name: str):
+        self.name = name
+
+    def maybe_append_choice(self, choices, **kwargs):
```
What is the "choices" argument here? ( e.g. datatype and intended usage). A small doc comment would help clarify I think.

kadeng commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,272 @@
+from ..cpp import CppPrinter, DTYPE_TO_CPP
+
+
+cexpr = CppPrinter().doprint
```
This might benefit from a type annotation and/or a small comment what it is, so we don't need to follow the link to CppPrinter when reading the code.

kadeng commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,272 @@
+    """
+    Kernels defined by the CUDA language.
+    """
+    overrides = OpOverrides
```
Which consequences does this line have? It's a bit hard to read out of the inductor codegen, how exactly these overrides fields are used.

kadeng commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,272 @@
+            return "0"
+        return str(node.get_layout().offset)
+
+    def ptr(self, node: IRNode, default_node: IRNode = None) -> str:
```
A small doc comment maybe?

ipiszy replied:
Added some comments.

kadeng commented on torch/_inductor/codegen/cuda/cuda_scheduling.py:
```diff
@@ -0,0 +1,46 @@
+from typing import List
+
+from ... import config
+from ...codecache import code_hash, get_path
```
It would be better to use absolute imports here, as that eases code navigation tools ( both in IDEs and github code views for example )

kadeng commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -0,0 +1,186 @@
+# import cutlass libs
+import scripts as cutlass_lib
+
+from ...autotune_process import CUDABenchmarkRequest, TensorMeta
```
Same here, absolute imports would be preferable for most tools

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @kadeng !

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_inductor/kernel/mm.py:
```diff
@@ -192,15 +179,16 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
-            prefix_args=1,
-            epilogue_fn=addmm_epilogue(layout.dtype, alpha, beta),
-        )
+    if use_triton_template(layout):
```
Add `m * n != 0` to the condition (now removed from the above)?

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented on torch/_inductor/scheduler.py:
```diff
@@ -1644,7 +1644,12 @@ def codegen(self):
+                if isinstance(node.node, ir.CUDATemplateBuffer):
+                    from .codegen.cuda.cuda_scheduling import CUDAScheduling
+
+                    CUDAScheduling(self).codegen_template(node, epilogue)
```
can_fuse() always returns False for CUDATemplateBuffer (in triton.py). It invokes Triton codegen_template() here to codegen only the non-epilogue part.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,309 @@
+
+        if node is None:
+            return None
+        return {**self.args.input_buffers, **self.args.output_buffers}.get(
```
OSS linter formats it in this way..

ipiszy reviewed (commented):
Thanks @aakhundov ! Updated the PR, ptal.

ipiszy commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,309 @@
+
+
+def _normalize_idx(index: int, total_length: int) -> int:
+    return index if index >= 0 else index + total_length
```
Actually I don't want to add assertions, as in `stride()` I just return a default value if index is out of bound (useful to get batch_size, batch_stride, etc, from a 2d matrix).

aakhundov replied:
Makes sense.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented on test/inductor/test_max_autotune.py:
```diff
@@ -1,323 +0,0 @@
-# Owner(s): ["module: inductor"]
```
No, it was a mistake ^^.

ipiszy commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2504,11 +2508,12 @@ def codegen_template(self, template_node, epilogue_nodes):
         # finalize must be called after adding epilogue above
-        src_code = partial_code.finalize()
+        # TODO: Maybe unify CUDATemplateKernel to also use PartialRender for flexible epilogue fusion.
+        src_code = partial_code if isinstance(partial_code, str) else partial_code.finalize()
```
Both Triton and CUDA execute this code path since CUDAScheduling inherits from TritonScheduling.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented:
@ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk.

aakhundov reviewed (changes requested):
@ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk. As the 3 PRs below this one in the stack are green and the one above is also red, my guess is that the root cause may be in this one.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
> @ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk. As the 3 PRs below this one in the stack are green and the one above is also red, my guess is that the root cause may be in this one.

Yes on it. There had been a bunch of test failures. I've fixed some and now there is something new. It's an iterative process..

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@pytorchbot label "topic: not user facing"

chenyang78 commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -0,0 +1,309 @@
+        input_reorder: The actual order of input nodes.
+                       e.g. The template might have input argument defined as [X, W, Bias],
+                       and the actual input passed into this template could be [Bias, X, W].
+                       In this case, the `input_reorder` would be [2, 0, 1].
```
Wondering why we allow different orders for the actual input and the template declaration in the first place? Could we enforce the actual input to have the same order specified by the template?

aakhundov replied:
I assume, for template front-end similarity with the existing Triton templates?

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 4: CUDA (template) kernels"

This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]
